### DISCUSSION TITLE
训练的模型，export后，识别能力变差

### DISCUSSION THREAD

**Discussion by mxdg** [PaddlePaddle/PaddleOCR#13958] [Q&A]
训练后，模型用infer_rec.py，识别正确，准确率99%：
python tools/infer_rec.py -c configs/my-ch_PP-OCRv4_rec_hgnet.yml -o Global.pretrained_model=./output/rec_ppocr_v4_hgnet/latest Global.infer_img=/root/autodl-tmp/jg3.png Global.use_gpu=false

导出模型，再去识别，内容识别错误，准确率66%：
python tools/export_model.py -c configs/my-ch_PP-OCRv4_rec_hgnet.yml -o Global.pretrained_model="./output/rec_ppocr_v4_hgnet/latest" Global.save_inference_dir="./output/v4_hgnet_1/"

导出有问题吗

**Comment by karanpate1** [comment]
> After training, the model uses infer_rec.py and recognizes correctly with an accuracy of 99%: python tools/infer_rec.py -c configs/my-ch_PP-OCRv4_rec_hgnet.yml -o Global.pretrained_model=./output/rec_ppocr_v4_hgnet/latest Global.infer_img=/root/autodl-tmp/jg3.png Global.use_gpu=false
> 
> Export the model and try it again. The content recognition error is 66% accuracy: python tools/export_model.py -c configs/my-ch_PP-OCRv4_rec_hgnet.yml -o Global.pretrained_model="./output/rec_ppocr_v4_hgnet/latest" Global.save_inference_dir="./output/v4_hgnet_1/"
> 
> Is there a problem with exporting?


###  Issue: High Validation Accuracy but Garbage Inference in PP-OCRv5 (Solved)

I fine-tuned **PP-OCRv5** on a custom dataset (English alphabets, numerals, and some symbols).

* Training accuracy: **98%**
* Validation accuracy: **99%**

But when I ran inference (even on images from training/validation sets), the predictions were totally wrong.

I double-checked everything:

* Followed official finetuning + export steps
* Tested with both `tools/infer_rec.py` and exported inference model
* Same problem: **terrible results**

At this point, I was stuck because evaluation showed **99% accuracy**, but inference was useless.

---

###  Root Cause

Turns out the problem was due to **mismatch in image preprocessing between training and inference**.

In my training config `PP-OCRv5_mobile_rec.yml`:

```yaml
sampler:
  name: MultiScaleSampler
  scales: [[320, 32], [320, 48], [320, 64]]
```

So during training, images were dynamically resized to **3 different shapes**:

* 3×32×320
* 3×48×320
* 3×64×320

But in my exported `inference.yml`, the preprocessing looked like this:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 48
      - 320
```

Notice the mismatch?

* Training: random **multi-scale** (32, 48, 64 heights)
* Inference: fixed to **48 height only**

That mismatch caused the network to completely fail at inference.

---

### ✅ Fix

I manually adjusted the `inference.yml` to try different scales from training.

* With **3×48×320**, results were trash.
* With **3×64×320**, also trash.
* With **3×32×320**, predictions became **perfect**, matching my training/validation accuracy.

So the fix was:
 **Set `RecResizeImg.image_shape` in `inference.yml` to the same shape(s) used in training.**

Final fix:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 32
      - 320
```
---

###  Takeaway

If you’re fine-tuning PP-OCR models with `MultiScaleSampler`, make sure your **inference preprocessing matches your training scales**. Otherwise, you’ll see high validation accuracy but garbage results at inference.

---

This should save people from wasting days like I did. 🙃

---

Want me to polish this into a proper **GitHub issue comment** format (with Markdown headings + code blocks, short and punchy)? Or do you want it more like a **blog-style post** with more storytelling?
